### ISSUE TITLE
collective_merge duplicates measurements

### ISSUE THREAD

**Issue by jgreitemann** [ALPSCore/ALPSCore#295]
This is probably not a bug but a matter of design. Calling `accumulator_set::collective_merge(comm, 0)` will currently collect the measurements from all MPI processes into the master, but also leave them in the slave processes. Calling `collective_merge` _again_ would thus give wrong results as the slave measurements would be duplicated but the master measurements wouldn't. I'd consider this a design flaw.

In a typical Monte Carlo simulation, this is not a problem since `collect_results` is only ever called once at the end of the simulation. However, we found this to be an issue when doing bold diagrammatic Monte Carlo where the results are collected occasionally to improve the bold approximation self-consistently but the accumulators are intentionally not reset but continue to accumulate data.

As a simple workaround, one can save the measurements to a checkpoint prior to the collective_merge and restore the master from that checkpoint after the merge. One can also keep a copy of the `accumulator_wrapper`s in memory (obtainable using the `clone` function, but not the copy ctor! -- Why?). Unfortunately, the latter has to be done for each accumulator separately; one cannot just copy the `accumulator_set` as a whole since its copy ctor doesn't do anything either. (Maybe this could also be addressed? In the code, it says, _TODO: how do we handle that?_ I fail to see why it couldn't just deep-copy the `wrapper_set`.)

The above workarounds do however copy the data unnecessarily. As an alternative, one could just call `accumulator_wrapper::reset()` on the slaves after a collective merge. I would argue that this should not be left up to the user, but should in fact be done in `collective_merge`.

This [example program](https://gist.github.com/jgreitemann/e9649178d90f78b07498b250c1364d06) demonstrates the issue.

**Comment by egull** [comment]
@galexv, can you give a status update? I just heard from Lode/Jonas that this one is critical for them.

**Comment by galexv** [comment]
@egull Ok, it has just went to my top-priority task list.

**Comment by mwallerb** [comment]
Just for mental bug-tracking for me: this is fixed in ALEA by commit d5c6c1e4 .
It is of course *not fixed* in the accumulators library.

**Comment by galexv** [comment]
So, the current version of the fix (commit as of 9bd2ff2) does not merge the autocorrelation lengths properly. But should it?

I am asking for the opinions regarding the sensibility of merging autocorrelation times (this question is also relevant to the new ALEA library).  The mean and the variance are properties of the measured random variable, while autocorrelation length (at least, as it's operationally defined now) seems to be a property of a particular Markov chain. What is a sensible way (if any) to combine the autocorrelation times of several Markov chains into a single value?

**Comment by jgreitemann** [comment]
Correct me if I'm wrong, but I'd say the autocorrelation time obtained from a naive binning analysis based on many independent Markov chains which have been concatenated remains valid, at least as far as it is used to estimate the real error of the accumulator. The whole point of the formation of successively larger bins is to decorrelate the bin averages. In the extreme case where the number of bins corresponds to the number of independent Markov chains, that decorrelation would in fact be perfect as the bins _are_ independent from each other and further coarse-graining would be pointless.

Thus, while I'd expect the autocorrelation time estimated for an accumulator which has been merged from many processes to be somewhat smaller than the autocorrelation time estimated for an accumulator with the same number of samples but drawn from a single Markov chain, that is actually correct because in the former case the samples are in fact less correlated. The catch here might be interpretability: the amount of update steps one has to wait in order to gain a "reasonably independent" (in the sense of autocorrelation diminished by a factor of e) configuration is of course the same in both cases. It's just that the autocorrelation function exhibits jumps, i.e. whenever the merged time series crosses over to the data from another chain, the exponential decay (with decay constant tau) of the autocorrelation function drops to zero immediately. From that picture, it is clear that this can only really affect the result if the autocorrelation is still sizable at those jumps. In that case, however, the estimate for the autocorrelation time is inherently too small anyway. For it to be correct, the binning analysis assumes that at the most coarse binning level the bins are essentially uncorrelated.

**TL;DR**: for practical purposes where the simulation (on each process or thread) spans many times the autocorrelation time, the naive binning analysis on the merged time series still gives the correct autocorrelation time. If that is not the case, the estimate for the autocorrelation time will be smaller, (correctly) reflecting the fact that the merged time series is less correlated compared to when the samples would have been obtained from a single Markov chain, but the interpretation of tau as the decay constant of the autocorrelation function is compromised.

**Comment by mwallerb** [comment]
@galexv, @jgreitemann   I agree with almost everything :-)

For `K>1` runs, If we understand `tau` as autocorrelation time, it is technically invalid, since, as you point out, you don't have a "thread of measurements" anymore but rather a nexus (common starting point) from which multiple threads emerge.  However, the definition of `tau_int` as correction factor to the law of large numbers still works, i.e., for a total number `N` of measurements over all `K` runs, we can define it s. th.:
```
                       Var[ Estimator_X ] = Var[X] * tau_int / N
```
holds.  This is also ultimately what we are interested in.

Now there are two ways forward:

 1. @jgreitemann 's strategy of "chaining" all the different threads together and performing binning analysis on that object.  This biases `tau_int` (by above definition) to lower values, since we introduce artificial "clearances" between the different chains.  That bias however becomes small for  `N >> K`.

 2. Assuming the individual runs to be independent, which translates to a simple weighed average over `tau_int`.  This biases `tau_int` to lower values in the case where the runs are not independent, and goes away as `N` becomes large.

Both things are okay I think :-)

**Comment by galexv** [comment]
@jgreitemann @mwallerb @egull ALEA (PR #357) will replace the current accumulators. How important is the correct autocorrelation after *repeated* merges for now --- do we need to have it fixed in the current code?

**Comment by jgreitemann** [comment]
From our end it's no matter of urgency. We [manually reset the accumulators](https://gitlab.lrz.de/Lode.Pollet/LecturesDiagrammaticMonteCarlo/blob/master/Froehlich_Selfenergy/bold_diagmc.cpp#L87) in the (more or less final) version of our bold DiagMC code. And for DiagMC specifically the autocorrelations are typically very small and the whole error analysis becomes somewhat meaningless in the self-consistent bold scheme anyway. It was the duplication of measurements that had us concerned.

As a matter of principle with regard to the implementation in ALEA, I'd like to point out again that if `N/K` is sufficiently small to bias the estimate `tau_int` as outlined above, that estimate is inherently biased anyway as bins on the most coarse level would still be correlated. Thus, I think the weighted average @mwallerb proposed is not really worth the hassle. In the end, the question is why one is interested in the autocorrelation time itself in the first place: is it to qualitatively judge if enough sweeps have been made for the binning analysis to become reliable (which will still be possible even if `tau` is biased and that bias will go away as the answer to that question becomes 'yes') or is the aim to quantitatively benchmark an update scheme and maybe even extract the dynamical exponent -- in this case additional scrutiny is needed anyway and one probably wouldn't attempted that within a parallel setting.

**Comment by mwallerb** [comment]
@jgreitemann 
> As a matter of principle with regard to the implementation in ALEA, I'd like to point out again that if N/K is sufficiently small to bias the estimate tau_int as outlined above, that estimate is inherently biased anyway as bins on the most coarse level would still be correlated.

Not if you thermalize for each run.

**Comment by galexv** [comment]
Ok, so, apart from the issue of autocorrelation, the accumulator is now [reset on ALPSCore side](https://github.com/ALPSCore/ALPSCore/blob/c987a08cf5b18d54149e164aecbb0d9feed67bec/accumulators/include/alps/accumulators/accumulator.hpp#L874) (PR #380).

(As a side-effect, there is no `const`-version of `collective_merge()` anymore).

**Comment by jgreitemann** [comment]
@mwallerb Well yes, but chances are that if you don't reach sufficiently many samples to decorrelate the Markov chain, either your thermalization phase is too short too, or thermalization takes most of the computation and it's pointless to parallelize in the first place.

**Comment by mwallerb** [comment]
@jgreitemann  Sure.  I think I will run a couple of experiments for an AR(1) model and see which strategy comes out on top.  My feeling is that an average over `tau_int` is somewhat more well-defined, but we'll see.

**Comment by egull** [comment]
OK, given this discussion I'll remove the 'bug' and 'ongoing' labels (we fixed this in PR #380) and assign it to Markus, who can take over for the new ALEA.
